src/extension.ts

```diff
@@ -107,18 +109,21 @@ function startServer(context: ExtensionContext, javaHome: string) {
   );
 
   context.subscriptions.push(
-    commands.registerCommand("metals.restartServer", async () => {
-      const serverPid = client["_serverProcess"].pid;
-      await exec(`kill ${serverPid}`);
-      const showLogsAction = "Show server logs";
-      const selectedAction = await window.showInformationMessage(
-        "Metals Language Server killed, it should restart in a few seconds",
-        showLogsAction
-      );
-
-      if (selectedAction === showLogsAction) {
-        client.outputChannel.show(true);
-      }
+    commands.registerCommand("metals.restartServer", () => {
+      // First try to gracefully shutdown the server with LSP `shutdown` and `exit`.
+      // If Metals doesn't respond within 4 seconds we kill the process.
+      const killTimeout = setTimeout(() => {
+        window.showWarningMessage(
+          "Metals is unresponsive, killing the process and starting a new server."
+        );
+        const serverPid = client["_serverProcess"].pid;
+        exec(`kill ${serverPid}`);
+      }, 4000);
+      client.sendRequest(ShutdownRequest.type).then(() => {
+        clearTimeout(killTimeout);
+        client.sendNotification(ExitNotification.type);
+        window.showInformationMessage("Metals is restarting");
+      });
```

Comment:
This look ok, I'll give it a try later.
Alternatively you can model this as follows:
- promise A, that fails after a timeout
- promise B, that tries to terminate gracefully
- Promise.race(a, b).catch(killPid)

But only if you think it's clearer

Reply:
Nice, that reads more cleanly. Merged your branch